home/vasco-santos/2021-04-exploration-report-fil-signing-tools.md

```diff
@@ -0,0 +1,66 @@
+- `@nodefactory/filsnap`
+  - Metamask snap (plugin) to enable Metamask users interaction with filecoin dapps.
+- `@glif/local-managed-provider`
+
+## Maturity
+
+Currently in `v.0.15.0`, this library is widely used, has been actively maintained over the last months and is the go to library for signing related operations in the ecosystem. It offers all the needed cryptographic primitives to interact with a Lotus node.
+
+## Project
+
+### Libraries
+
+1. Rust Native Library
+  - Secp256k1
+  - BLS
+  - Filecoin transactions (CBOR <> JSON serialization)
+  - Multisig (in progress)
+2. WASM Library (Browser and Node.js)
+  - Secp256k1
+  - BLS
+  - Filecoin transactions (CBOR <> JSON serialization)
+  - Multisig (in progress)
+3. JSON RPC Server
+  - Exposes most of the functions available in the signing library
+
+### Cryptographic primitives
+
+1. `generateMnemonic`
+2. `keyDerive`
+3. `keyDeriveFromSeed`
+4. `keyRecover`
+5. `transactionSerialize`
+6. `transactionSerializeRaw`
+7. `transactionParse`
+8. `transactionSign`
+9. `transactionSignLotus`
+10. `transactionSignRaw`
+11. `verifySignature`
+12. `addressAsBytes`
+13.  `bytesToAddress`
+
+All available via Rust, Wasm (browser and Node.js) and JSON RPC Server.
+
+## Key Findings
+
+- Per [https://github.com/Zondax/filecoin-signing-tools/issues/373](https://github.com/Zondax/filecoin-signing-tools/issues/373), [Zondax/filecoin-signing-tools#358](https://github.com/Zondax/filecoin-signing-tools/issues/358), [Zondax/filecoin-signing-tools#356](https://github.com/Zondax/filecoin-signing-tools/issues/356) and [Zondax/filecoin-signing-tools#352](https://github.com/Zondax/filecoin-signing-tools/issues/352), there are a few issues with bundlers and browsers.
+- Per [Zondax/filecoin-signing-tools#368](https://github.com/Zondax/filecoin-signing-tools/issues/368) and [Zondax/filecoin-signing-tools#343](https://github.com/Zondax/filecoin-signing-tools/issues/343), there are several issues with WASM.  The browser WASM example uses the pure JS implementation and an issue tracking how wasm is being loaded in the browser. Could not find any project leveraging the WASM library. Is it working?
+
+## Overall Recommendations
+
+- The [documentation site](https://zondax.ch/updating-documentation) as the main source for docs should be available. It has been on maintenance, which makes it difficult to understand what methods are supported unless we check the code/tests and this should be handled.
+- Clarify WASM library status per above findings
+- Friction onboarding
+  - Several examples are provided, but some are not easy to run and do not have instructions. It would be extremely valuable for the community to have instructions for running this examples together with some educational content regarding what are the use cases for each and what happens under the hood.
+
+## Potential Action Items for w3dt
+
+- Run the WASM code in the browser and compare the bundle size with the pure JS implementation
```

Comment:
Any thoughts on why there's the two implementations? It sounds like the WASM one isn't even being used (by that issue that points out that even the example is using the JS implementation!). Maybe WASM was there for a reference before the JS implementation.
The reason for the question is that there may be something else to measure here -- is it a perf thing perhaps? If it's simple enough, maybe a task here is to purify the JS implementation, stripping away the WASM bits, make it more bundlable and friendly for consumption.

Reply:
They have this structure: https://github.com/Zondax/filecoin-signing-tools/tree/c0177f73efd089f60c01d88adc22926006fd693d/signer-npm#structure

Which gives me the impression that they want to ship both versions (JS and wasm), but there is no `pkg` folder, nor example apparently using it. 

I was thinking on a different perspective. Given that this bundles cryptographic primitives, I would like to see what are the difference between using JS or Wasm in terms of bundle size, so that we can take learnings for libp2p's crypto.

Do you know of someone we can reach out to get to know more?

> make it more bundlable and friendly for consumption

This definitely should happen.

Reply:
Yeah, it does seem like not everything is in GitHub that you need to reproduce the published package(s), it's an odd setup I can't quite figure out.

Digging into it a bit further, the package contents are quite interesting https://unpkg.com/browse/@zondax/filecoin-signing-tools@0.15.0/ - note package.json points to a different entry point for browser vs node but both paths seem to have the same .wasm but load them with different wrappers. I wonder if these are products of the Rust compile toolchain?

I'll get us a better venue to chat with Zondax about this and see what we can learn.